feat(async-flow): asyncFlow

erights opened the pull request:
closes: #9302 
refs: #9125, #9126 #9153 #9154, #9280

## Description

Upgrade while suspended at `await` points! Uses membrane to log and replay everything that happened before each upgrade.

In the first incarnation, somewhere, using a ***closed*** async function argument
```js
const wrapperFunc = asyncFlow(
  zone, 
  'funcName`, 
  async (...) => {... await ...; ...},
);
```
then elsewhere, as often as you'd like
```js
const outcomeVow = wrapperFunc(...);
```

For all these `asyncFlow` calls that happened in the first incarnation, in the first crank of all later incarnations
```js
asyncFlow(
  zone, 
  'funcName`, 
  async (...) => {... await ...; ...},
);
```
with async functions that reproduce the original's logged behavior. In these later incarnations, you only need to capture the returned `wrapperFunc` if you want to create new activations. Regardless, the old activations continue.

#### Future Growth

I designed this PR so it could grow in the following ways:

- TODO: The membrane should use the `HandledPromise` API to make proper remote presences and handled promises, so that the guest function can use `E` on objects or promises it receives from the host as expected. I commented out the additional ops needed for these: `doSend` and `checkSend`.

- TODO: Currently, I assume that the host side only presents vows, not promises. However, imported remote promises can be stored durably, and be durably watched, so the membrane should be extended to handle those.

- TODO: We currently impose the restriction that the guest cannot export to the host guest-created remotables or promises. (It can pass back to the host remotables or promises it received from the host.) I commented out the additional ops needed for these: `doCall`, `checkReturn` and `checkThrow`. I wrote the `bijection` and `equate` subsystems so that old durable host wrappers can be hooked back up on replay to the corresponding new guest remotables and promises.

### Security Considerations

Nothing enforces that the argument async function is closed, i.e., does not capture (lexically "close over") any direct access to mutable state or ability to cause effects. If it does, it still cannot harm anything but itself. But it -- or its replayings -- may be more confusable, and so more vulnerable to confusion attacks.

Since this entire framework itself is written as a helper (well, a huge helper) with no special privilege, it cannot be used to do anything that could not have otherwise been done. It is not a source of new authority.

See caveats in Upgrade Considerations about isolation of effects following a replay failure.

### Scaling Considerations

We assume that the total number of async functions, i.e., calls to `asyncFlow`, are low cardinality. This is essential to the design, in exactly the same sense as our assumption that exoClasses are low cardinality. The RAM cost is proportional to the number of these.

The current implementation further assumes that the total number of activations of these replayable async functions are low cardinality. This will likely be a scaling problem at some point, but is unlikely to be painful for the initial expected use cases.

The current implementation imposes two simplifying restrictions that allow us to relieve some of this memory pressure: Currently, the async function argument cannot make and export new remotables or promises to the host. Thus, once its outcomeVow is settled, its job is done. There is very little more it can do that is observable. Thus, once this outcome is settled, the activation is shut down and most of the memory it held onto is dropped.

Of the activations not shut down, they must replay from the beginning in each incarnation. If a given activation has a long history of past activity, this can become expensive.

How do we verify in CI that when an asyncFlow is in use & when it has completed, resource usage in RAM & on disk meet our expectations?

The PR assumes `low cardinality` of asyncFlows. what scale is `low cardinality` - Is 10^3, 10&5? What is the risk if cardinality is too high?

### Documentation Considerations

For the normal developer documentation, `asyncFlow` should make things simpler and more like "just JavaScript". The membrane translates between host vows and guest promises, so the async function can simply `await` on promises without needing the `when` from `@agoric/vow`.

### Testing Considerations

This PR is structured as a tower of building blocks, where I unit tested each as I went, in bottom up order, in order to build with confidence. Currently, each of these building blocks is also very paranoid about internal consistency checking, so I'd get early indications if I made a mistake. Probably some of this internal consistency checking can be reduced over time, as we gain more static confidence.

This PR is currently using the fake upgrade testing framework from the `@agoric/zone` package. This caused bug #9126. Instead, we need to redo all these tests with a real upgrade testing framework, like the one in bootstrapTests. See https://github.com/Agoric/agoric-sdk/blob/master/packages/boot/test/bootstrapTests/test-zcf-upgrade.ts


### Upgrade Considerations

The point.

In an reviving incarnation, if the async function argument of
```js
asyncFlow(
  zone, 
  'funcName`, 
  async (...) => {... await ...; ...},
);
```
fails to recapitulate the logs of each of its activations, those activations will not have done any damage. They will simply be stuck, with a diagnostic available via 
```js
adminAsyncFlow.getFailures(),
```
To unstick these, so those stuck activations can continue to make progress, upgrade again using an async function argument that does reproduce the logged behavior of each of its activations.

#### Caveat: Imperfect isolation of effects following a replay failure

Once a replay failure is detected, we attempt to isolate the replaying activation from its outside world, and to also shut down its further execution as soon as possible. But it may be in the midst of activity with a non-empty stack. Following a replay failure, we have no way to preemptively terminate it without any further execution of the activation. This further execution may therefore be arbitrarily confused. We simply try to isolate it as much as possible, immediately revoking all access it had through the membrane to all authority to cause observable effects. However,
- We do not consider `console` logging activity to be an observable effect. These might be caused by diagnostics emitted by this framework in response to its "isolated" confused behavior.
- Because we do not consider `console` logging to be an observable effect, we also allow this as an exception to our closed function rule. Messages it sends directly to the console are not logged, and can differ without causing replay failure. During its post-replay-failure confused execution, it can still directly log to the console.
- It is not resource limited, so its post-replay confused execution can accidentally engage in resource exhaustion attacks, including infinite loops. However, the vat as a whole is resource limited. An infinite loop will eventually crash the vat, which can then be recovered with yet another upgrade.
   - Because of metering, an activation that executed successfully in a previous incarnation might not replay correctly, even if it doesn't cause any explicit side-effects. That's because metering is a hidden side-effect of any execution.

erights commented on packages/vow/src/index.js:
```diff
@@ -1,6 +1,7 @@
 // @ts-check
 export * from './tools.js';
 export { default as makeE } from './E.js';
+export { VowShape } from './vow-utils.js';
```
export isVow too

erights commented:
OMG it's ready for review!

ivanlei commented:
additional `Scaling Considerations`
1. how do we verify in CI that when an asyncFlow is in use & when it has completed, resource usage in RAM & on disk meet our expectations?

erights commented:
> additional `Scaling Considerations`
> 
> 1. how do we verify in CI that when an asyncFlow is in use & when it has completed, resource usage in RAM & on disk meet our expectations?
> 2. the PR assumes `low cardinality` of asyncFlows. what scale is `low cardinality` - Is 10^3, 10&5? What is the risk if cardinality is too high?

Added to the Scaling Considerations of the PR comment.

erights commented on packages/async-flow/package.json:
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "@agoric/async-flow",
+  "version": "0.0.1",
```
What is our standard practice for a first version number?

erights replied:
Done

erights commented:
PR comment said:

> ~As a rapid prototyping strategy, I'm doing all this in `src/async-flow/` and `test/async-flow` subdirectories of `@agoric/zone`. But the plan is to move all these new files to a separate `@agoric/async-flow` package and restore `@agoric/zone` to its pristine state -- essentially the state before this PR.
Part of that restoration will be to remove all unneeded additions to Zone's `package.json`.~

I have now moved all that new stuff to a new @agoric/async-flow package and removed all mods to @agoric/zone.

Reviewers, I don't understand all the config that goes into setting up a new agoric-sdk package. I just cargo culted it from @agoric/zone. Please double check that I got the new package setup correct.

michaelfig commented:
> 2. the PR assumes `low cardinality` of asyncFlows. what scale is `low cardinality` - Is 10^3, 10&5? What is the risk if cardinality is too high?

I think there should be some clarification around this scaling consideration.  `asyncFlow(...)` returns a wrapper function. 
 The "low cardinality" does not refer to calls to (activations of) a given wrapper function, only to individual calls to `asyncFlow` itself, each with a unique `funcName`.  This is the same sort of concern that defining a new `exoClass` should be "low cardinality", but there is no such caveat on the number of *instances* of a particular `exoClass` that can be created.

erights commented:
> > 2. the PR assumes `low cardinality` of asyncFlows. what scale is `low cardinality` - Is 10^3, 10&5? What is the risk if cardinality is too high?
> 
> I think there should be some clarification around this scaling consideration. `asyncFlow(...)` returns a wrapper function. The "low cardinality" does not refer to calls to (activations of) a given wrapper function, only to individual calls to `asyncFlow` itself, each with a unique `funcName`. This is the same sort of concern that defining a new `exoClass` should be "low cardinality", but there is no such caveat on the number of _instances_ of a particular `exoClass` that can be created.

Hi @michaelfig , there is both. Certainly, a prepared guest function is like an exoClass (and is internally implemented by an exoClass). And guest func activations are like exoClass instances (and are internally implemented...). But an *active activation* is inherently an guest function in motion, with all the attendant ephemeral state. So the best we can do is to limit further heap-only state to be proportional to the number of active activations. So let's distinguish the lifecycle stages of an guest function activation. I am making up these names now. Bikeshedding better names is welcome.

- ***active current***. This is the state they're born into as of the first invocation that creates and starts the activation. When there is an actual JS guest `activation paused at an `await`, it is necessarily of this form.

- ***sleeping***. If in incarnation i1 the activation was paused waiting on a vow to settle, and in incarnation i2 the vow has not yet settled, then *ideally* we can leave the durable replay log and other state in durable storage. We should not need to reactivate it until that vow actually settles. 

- ***active replaying***. Once a vow it was `await`ing on in a previous incarnation does settle, we need to restore the guest activation to a heap manifestation as a JS `async` and set it in motion replaying from its durable log. Once it finishes replaying from the log, it has caught up and transitions to ***active current***, reaching the `await` at which it slept. If the promise it is `await`ing corresponds to a vow that's now settled, it continues as ***active current*** from that point.

- ***replay failure***. If during the ***active replaying*** state the guest activation fails to exactly reproduce its previously logged behavior, it becomes inactive as ***replay failure*** associated with a diagnostic explaining how the replay failed, so it can be repaired in another future upgrade. As of the next incarnation, the failure status is cleared and all these go into ***active replaying*** in order to retry replaying the from the log.

- ***complete***. If the promise the guest activation returned settles, we assume that the job of the guest activation is done. It then goes into a durably done state, dropping all its bookkeeping beyond just remembering that it is done. The replay logs and membrane state are all dropped.

We necessarily maintain heap state proportional to the number of activations in the ***active current*** and ***active replaying*** states. Further, for each of these, we necessarily maintain heap state proportional to the number of logged events across the membrane, and the number of remotables and promises sent through the membrane. No choice about these. The technique cannot work with less. (Well, not without a lot of pain.)

I don't believe we have heap state associated with the ***sleeping*** state. However, the current implementation does not let sleeping activations lie. Rather, as a current implementation restriction, it eagerly wakes these up into ***active replaying***. But the code is designed so we can fix this, waking them up in a lazier manner.

There should be no heap storage proportional to the number of ***complete*** guest activations.

None of these claims are tested.

erights commented:
![image](https://github.com/Agoric/agoric-sdk/assets/273868/1a6b3998-cb0e-4dcd-b5ff-79a9f392a6d2)

michaelfig commented on packages/async-flow/test/test-convert.js:
```diff
@@ -0,0 +1,118 @@
+  testConvert(t, zone1, vowTools1);
+
+  // These converters keep their state only in the bijection,
+  // which looses all its memory between incarnations.
```
```suggestion
  // which loses all its memory between incarnations.
```

erights replied:
Done.

michaelfig commented on packages/async-flow/test/test-equate.js:
```diff
@@ -0,0 +1,121 @@
+  testEquate(t, zone1, vowTools1);
+
+  // equate keeps its state only in the bijection,
+  // which looses all its memory between incarnations.
```
```suggestion
  // which loses all its memory between incarnations.
```

erights replied:
Done

michaelfig commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,402 @@
+            // the first call to a host object.
+            membrane.wake();
+
+            // We do *not* call the guesAsyncFunc by having the membrane make
```
```suggestion
            // We do *not* call the guestAsyncFunc by having the membrane make
```

erights replied:
Done

michaelfig commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,402 @@
+            // the guest presence of other host objects.
+            //
+            // `bijection.hasGuest(guestResultP)` can be false in a delayed
+            // guest - to - host setlling from a previous run.
```
```suggestion
            // guest - to - host settling from a previous run.
```

erights replied:
Done

michaelfig commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,402 @@
+            // having any observable effects on the host, aside from
+            // console logging and
+            // resource exhaustion, including infinite loops
+            return undefined;
```
Can you make the return result describe this situation better?  Maybe something like:
```suggestion
            return `panic: see getFailures() for more information`;
```

erights replied:
Even better: made it a `throw Fail`. Done.

michaelfig reviewed (commented):
The code looks pretty good, but I'm still a little hesitant to give official approval.  I think it's important for others to weigh in as well.

erights commented on packages/async-flow/src/convert.js:
```diff
@@ -0,0 +1,120 @@
+  const convert = (specimen, label) => {
+    // Open code the synchronous part of applyLabelingError, because
+    // we need to preserve returned promise identity.
+    // TODO switch to Richard Gibson's suggestion for a better way
+    // to keep track of the error labeling.
```
IIRC, it is an explicit stack of labels pushed and popped from an internal mutable array. @gibson042 , could you correct or clarify my explanation and provide a link? Thanks.

gibson042 replied:
https://github.com/endojs/endo/pull/1795#issuecomment-1756093032

I'll be revisiting it after introducing `esbench` for microbenchmarking.

erights replied:
Great! In that case, resolving here as any such change can happen in both places after this PR.

erights commented on packages/async-flow/src/log-store.js:
```diff
@@ -0,0 +1,139 @@
+        return result;
+      },
+      pushEntry(entry) {
+        const { state, self } = this;
+        !self.isReplaying() ||
+          Fail`still replaying: ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        tmp.for(self).index === state.mapStore.getSize() ||
+          Fail`internal: index confusion ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        state.mapStore.init(tmp.for(self).index, entry);
+        tmp.for(self).index += 1;
+        tmp.for(self).index === state.mapStore.getSize() ||
+          Fail`internal: index confusion ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        return tmp.for(self).index;
```
I used to write code like 
```js
const { state: { mapStore }, self } = this;
```

but `yarn format` turns that into
```js
        const {
          state: { mapStore },
          self,
        } = this;
```

so I took a suggestion from @turadg and only extract one level `{ state, self }` or `{ state, facets }` on that first line, and then work from there. But your other suggestions for reducing redundancy look good.

erights replied:
Much better, thanks. Done.

erights commented on packages/async-flow/src/weak-bijection.js:
```diff
@@ -0,0 +1,128 @@
+        !h2g.for(self).has(h) ||
+          Fail`internal: unexpected h->g ${h} -> ${g2h.for(self).get(h)}`;
+        return false;
+      }
```
Much better, thanks. Done.

erights commented on packages/async-flow/test/prepare-test-env-ava.js:
```diff
@@ -0,0 +1,30 @@
+export const asyncFlowVerbose = () => {
+  // TODO figure out how we really want to control this
+  // But keep in mind that all this async-flow stuff will migrate
+  // into its own package
```
Thanks for catching. Done.

erights commented:
Since this is closing in, and is still independent of https://github.com/Agoric/agoric-sdk/pull/9153 which is not yet reviewed, I unstaged this from #9153 so it can merge when ready.

erights commented on packages/async-flow/src/type-guards.js:
```diff
@@ -0,0 +1,48 @@
+
+export const LogEntryShape = M.or(
+  // ////////////////////////////// From Host to Guest /////////////////////////
+  ['doFulfill', HostVowShape, M.any()],
+  ['doReject', HostVowShape, M.any()],
+  // [
+  //   'doCall',
+  //   M.remotable('host wrapper of guest target'),
+  //   M.opt(PropertyKeyShape),
+  //   M.arrayOf(M.any()),
+  //   M.number(),
+  // ],
+  // [
+  //   'doSend',
+  //   M.or(M.remotable('host wrapper of guest target'), VowShape),
+  //   M.opt(PropertyKeyShape),
+  //   M.arrayOf(M.any()),
+  //   M.number(),
+  // ],
```
Already implemented ;) 
```js
      const guestEntry = harden([
        'checkCall',
        guestTarget,
        optVerb,
        guestArgs,
        callIndex,
      ]);
```
where the `guestArgs` get hardened.

Once hardened, everything passing through the membrane must indeed be passable.

Done.

erights commented:
Reviewers,

At https://github.com/Agoric/agoric-sdk/pull/9097#pullrequestreview-1964889488 @michaelfig said
> The code looks pretty good, but I'm still a little hesitant to give official approval. I think it's important for others to weigh in as well.

I have since dealt with all of @michaelfig 's review suggestions. It is up to the rest of you to help me make progress towards approval. Please let me know what would help. An interactive walkthrough with y'all?

erights commented:
In case it helps the review process, I did break create two draft PRs that sum to this one, where each is about one half and the layering helps understanding: 
- https://github.com/Agoric/agoric-sdk/pull/9175
- https://github.com/Agoric/agoric-sdk/pull/9176

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+        },
+      },
+    );
+    const makeAsyncFlowKit = (activationThis, activationArgs) => {
+      const asyncFlowKit = internalMakeAsyncFlowKit(
+        activationThis,
+        activationArgs,
+      );
+      asyncFlowKit.flow.restart();
+      return asyncFlowKit;
+    };
```
Why is this needed? The caller `asyncFlow` is "internal", right? Or is `prepareAsyncFlowKit` meant to be used outside of tests?

erights replied:
> Or is `prepareAsyncFlowKit` meant to be used outside of tests?

Yes. The `asyncFlow` function is a convenience we expect to account for typical usage. But it weirdly breaks our normal `prepareFoo` patterns. For those wishing to more directly abstract of async flow --- like perhaps orchestration! --- we also expose the normal `prepareFoo` APIs.

Resolving.

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+  // Cannot call this until everything is prepared
+  // adminAsyncFlow.wakeAll();
+
+  return harden({
+    prepareAsyncFlowKit,
+    asyncFlow,
+    adminAsyncFlow,
+  });
```
Should this return the prepared `vowTools`, so it can be used as the "vow entrypoint"?

erights replied:
Let's revisit this once we're addressing https://github.com/Agoric/agoric-sdk/issues/9231 in later PRs.

Resolving

mhofman reviewed (commented):
This is a partial review / preliminary questions. I've only skimmed through the higher level bits of asyncFlow for now

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+              // Extra paranoid because we're getting
+              // "promise watcher must be a virtual object"
+              // in the general vicinity.
+              zone.isStorable(vowish) ||
+                Fail`vowish must be storable in this zone (usually, must be durable): ${vowish}`;
```
As discussed, a promise cannot be stored durably, even if it is imported / externally decided. While I expect that we'll deal mostly with host promises that resolve "immediately" to a vow, we can't assume that. As such we should "wrap" any host promises into a locally created vow, which is resolved by using `watch` on the host promise.

erights replied:
Per our slack conversation (which we should capture), I will now insist on only vow on the host side, rejecting promises rather than autocoercing them. Seems less user friendly until we get to the diagnostics needed around perverse promise cases. Better to just leave this coercion for the users to do manually. At least as if this first PR. Resolving.

mhofman reviewed (commented):
Couple notes from our discussion earlier

erights commented on packages/async-flow/docs/async-flow-states.md:
```diff
@@ -0,0 +1,15 @@
+
+- ***replay failure***. If during the ***active replay*** state the guest activation fails to exactly reproduce its previously logged behavior, it becomes inactive as ***replay failure*** associated with a diagnostic explaining how the replay failed, so it can be repaired by another future upgrade. As of the next incarnation, the failure status is cleared and all these go into ***active replay*** ...
+
+- ***done***. If the promise the guest activation returned settles, we assume that the job of the guest activation is done. It then goes into a durably done state, dropping all its bookkeeping beyond just remembering that it is done. The replay logs and membrane state are all dropped.
```
There can be. We do expose the kit facets

erights replied:
I think this addresses the intent of the question, so I'm resolving this conversation. But I'm not sure. If not, please reopen and clarify. Thx.

erights commented:
Reviewers, do any of you mind if I squash the current tower of 6 commits into one? Is it ok?

mhofman commented:
> Reviewers, do any of you mind if I squash the current tower of 6 commits into one? Is it ok?

Fine by me

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+ */
+export const prepareAsyncFlowTools = (outerZone, outerOptions = {}) => {
+  const {
+    vowTools = prepareWatchableVowTools(outerZone),
```
I explain the underlying issue at https://github.com/Agoric/agoric-sdk/issues/9231 , and added some TODO comments citing that issue. Having done that, I'm closing this conversation, leaving that issue as something to be addressed after this PR. If this is not adequate for this PR, please reopen. Thx.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+  const prepareAsyncFlowKit = (
+    zone,
+    tag,
+    optGuestAsyncFunc = undefined,
```
It doesn't. Gone.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+              throw Fail`Function must either be in prepareAsyncFlowKit or restart args: ${q(
+                tag,
+              )}`;
+            }
```
Gone.

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,389 @@
+        outcome = performCall(...args);
+      }
+    } catch (fatalError) {
+      return panic(fatalError);
```
> I'm actually unsure either is right. Really what we want is to return something that will prevent the guest from making any further forward progress if possible.

IIUC, there is nothing we could return or throw that would prevent the guest from making further progress. If I'm missing something, please reopen.

I used to have `panic` return `undefined` but now I do have it always throw. The `return` above was vestigial. Fixed.

The best we can do with post-revoke behavior of the guest code is to prevent it from doing any damage, under the assumption that it was not written to do any damage. AFAICT, the best we can do is to address the following issues, which are good enough when the guest is assumed not to be attacking the host
- Even a benign guest should be assumed to be written naively to execution after revocation. Thus, once revoked, a naive guest may "go crazy", acting in ways outside its author's expectations. However, JS code is generally written assuming any call might throw rather than return. Now that a revoked membrane always throws, this should restore the assumptions made by the author of even a naive guest function.
- If the author has not thought about the throw case, then they probably do not catch the throw but let it propagate, thus effectively terminating the guest execution immediately. If they do catch it with a `try/catch`, then they probably have put some thought into being throw-tolerant.
- The membrane prevents any integrity effects from guest to host, effectively isolating the guest re integrity. For these purposes, there is no confidentiality concern. The remaining "go crazy" issue would be unreasonable resource usage, such as an infinite loop.

If I've missed any remaining dangers for post-revocation guest functions, please reopen and explain. Attn @raphdev @LuqiPan . Thanks. Resolving

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,389 @@
+      }
+    } catch (fatalError) {
+      return panic(fatalError);
+    }
```
Note that the quoted line has already been changed to `throw panic(fatalError)` as you previously suggested.

> This treats internal errors the same as replay errors?

Yes. I don't see any better option.

> I guess that putting a flow in a failure mode is fairly safe, even if the guest function has nothing to do with it.

Yes. And doing anything else would be less safe ;)

Resolving.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+
+            state.isDone = false;
+          },
+          done() {
```
Just answering this part first

> but I guess vows don't have an identity managed by liveslots.

With @michaelfig 's approval, we are using the remotable within the vow as the vow key, even though it is kinda an accidental part of the representation. I don't see any other reasonable alternative.

@michaelfig , to encapsulate this issue, should the function abstracting a vow to a key for the vow be exported from the vow package, to encapsulate the representational accident?

michaelfig replied:
> to encapsulate this issue, should the function abstracting a vow to a key for the vow be exported from the vow package, to encapsulate the representational accident?

Yes, I'd like that.

erights replied:
> Yes, I'd like that.

That part done.

> That means removed from eagerWakers if applicable

Done

> and from flowForOutcomeVowKey

Done

> which makes it interesting

was straightforward ;)

raphdev commented:
I see no issues with the code as is. I appreciate the thought put into the code and its considerations.

My main concerns, are that many of the scaling and performance considerations hinge on claims and assumptions that are untested and difficult to confidently evaluate about from source review alone. Without testing these claims, we have no information on performance characteristics and impact of resource usage; it's not clear to me current tests are comparable to usual-case scenarios. It would be ideal and beneficial to this feature to see these assertions tested:


> The RAM cost is proportional to the number of these.

> the total number of activations of these replayable async functions are low cardinality. This will likely be a scaling problem at some point, but is unlikely to be painful for the initial expected use cases.

> How do we verify in CI that when an asyncFlow is in use & when it has completed, resource usage in RAM & on disk meet our expectations?

It would be good to have tests or benchmarks which prove this meets our requirements and they aren't indeed painful scaling-wise today. 

In particular, we want to be sure

> once this outcome is settled, the activation is shut down and most of the memory it held onto is dropped.

Any lingering memory would impact performance and sustainability.

> If a given activation has a long history of past activity, this can become expensive.

It would be good to understand how expensive, and how long of a history we are able to support before we see adverse effects.

As far as I can tell, some of the potentially higher impact risks are in performance/scalability.

Without validating these assumptions before shipping there is potential risk we may see issues with real-world immediate or future usage. How do we gain certainty with these assumptions to avoid surprises?

raphdev commented:
Not a requirement to merge but something previously discussed around closed function that would be DX improvements for users:

* Documenting this caveat
* A lint rule that warns on non-closed function with `asyncFlow`

mhofman commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,389 @@
+        name,
+        makeGuestMethod(name),
+      ]);
+      // TODO in order to support E *well*,
+      // use HandledPromise to make gRem a remote presence for hRem
```
The problem with not implementing E right away is that any usage of eventual-send on a remote presence will currently result in an exception thrown, and when we update the asyncFlow helper, it will become a success, with an observable log difference, so any in-progress flow invocation will fail to replay unless extra steps are taken.

erights replied:
Agreed that we cannot commit to any log state until this is fixed. We're about to work on this, but I think I still want to do it in a separate PR staged on this one. So, resolving.

mhofman commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,389 @@
+    void when(
+      hVow,
+      hostFulfillment => {
+        if (!log.isReplaying() && guestPromiseMap.get(promise) !== 'settled') {
```
In what case would we have a guest already settled when we're not under replay? I can only imagine if somehow the host vow fulfilled twice, which would be erroneous behavior from the vow.

@erights you tried to remove these checks, which resulted in a behavioral difference and test failures. We should investigate why.

erights replied:
The test is checking whether the promise was already resolved. When I turn it off
```suggestion
        if (!log.isReplaying() /* && guestPromiseMap.get(promise) !== 'settled' */) {
```
I get an extra log record
```js
    ['doFulfill', v1, 'x'],
```
at index 10 in the log that is redundant with the same
```js
    ['doFulfill', v1, 'x'],
```
at index 0. So the behavior with the check present is the correct behavior. 

The question is, why is it happening twice? I've managed to falsify every hypothesis I came up with. Gonna keep looking, but might need help.

erights replied:
I got it. When the guest promise is created, it registers with the vow to get fulfilled once the vow is fulfilled. On replay, the vow is fulfilled, so when the guest promise for that vow is created, it registers. However, replay first fulfills that guest promise as instructed by interpreting the log's doFulfill. The but the `when` that registered the callback when the promise is created will still fire. The extra test is suppress reaction to the `when` callback.

So, regarding this issue, I'd say the code is correct as is. 

But walking through this makes me worried about the opposite case: If the vow is fulfilled, and the callback called, while we're replaying a log that has not recorded that this vow was settled. I think what we need to do is buffer those settling notifications and then react to them after replay is done. I think this is difficult to arrange because replay happens without really pausing. But each time the guest function does `await`s, it is an interleaving point. The callback might fire at that time, and the `!log.isReplaying()` above will cause that notification to get lost.

So, leaving this open until I either verify or falsify this opposite problem.

erights replied:
Problem verified and fixed. I verified that the new test case at `test-replay-membrane-settlement.js` fails before the fix and passes after the fix.

The fix was easy. Testing it turned out to be surprising hard to figure out, but easy enough to write once figured out.

Done!

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,410 @@
+            );
+            const eph = tmp.for(flow);
+            eph.membrane = membrane;
+            const guestThis = membrane.hostToGuest(activationThis);
```
From our conversation: let's make `this` inside the guest be a wrapper for the async flow itself. Passing the async flow instance to the guest allows it to reference the current invocation, which in the future would allow us to introduce features that lets an upgraded flow conditionally execute new behavior (e.g. ask question like "am I at a point where I'm still replaying") without violating the replay log (by calling a power outside of the membrane referencing the flow instance).

`asyncFlow` itself was never meant to support exo methods directly (the `this` context is not passable), and would always have required another wrapper on top, which can use the args just as much as the receiver to plumb the exo context through the membrane.

erights replied:
Leaving this one open and undone for now. I suspect that we'll want to expose a new facet designed for this purpose, rather than exposing the existing `flow` facet.

erights replied:
How do you feel about not making any such change in this PR, leaving this design issue for following PRs to address?

mhofman replied:
At the very least I would like `this` to be undefined for now so no code starts relying on it.

erights replied:
Done

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+              failures.delete(flow);
+            }
+
+            !state.isDone ||
+              // separate line so I can set a breakpoint
+              Fail`Cannot restart a done flow ${flow}`;
+
```
It is indeed. Gone.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+            const { outcomeKit } = state;
+            const { flow } = facets;
+
+            flow.wake();
```
The intuition was lazy evaluation. If someone wants the outcome, then if we were asleep, we should wake up so we can give them the outcome if we can. However, since this is a low level API and the `wake` method is exposed, I took your suggestion. Gone.

mhofman replied:
I did not follow the intuition: if the flow is sleeping, that means it's pending on some external input. waking it will simply change its state, but not affect the outcome.

Anyway, since it's gone, no need for any follow-up.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+            if (eagerWakers.has(flow)) {
+              // For now, once an eagerWaker, always an eagerWaker
+              // eagerWakers.delete(flow);
+            }
```
Good point. Moved it to the `done` method uncommented. Done.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+    const wrapperFunc = {
+      [hostFuncName](...args) {
+        const { flow } = makeAsyncFlowKit(this, args);
+        const outcomeVow = flow.getOutcome();
```
`getOutcome` waking it is now gone.

Resolving. But if you had other extraneous operations in mind, please unresolve.

erights commented on packages/async-flow/src/weak-bijection.js:
```diff
@@ -0,0 +1,144 @@
+});
+
+/**
+ * Makes a store like a WeakMapStore except that Promises and Vows can also be
+ * used as keys.
+ * NOTE: This depends on promise identity being stable!
```
> There is a big gotcha though: a "recognizable" only vow will not be collected.

I don't think it is possible to do any better, so resolving.

But attn @raphdev , as this is likely within the scalability issues you're examining.

> In general it feels really dangerous to rely on promise identity. I've actually been wanting to change liveslots to create new promise instances everytime a promise is unmarshalled.

Now that we're rejecting host-side promises, the only issue is guest-side promises. And the only guest-side promises we accept are guest promise wrappers for host vows, so I think this problem disappears until we allow promises in more places. Still resolving.

erights commented on packages/vow/src/vow-utils.js:
```diff
@@ -38,3 +38,14 @@ export const getVowPayload = specimen => {
 };
 harden(getVowPayload);
+
+export const vowishKey = k => {
```
Should probably find a better name

erights replied:
All clear names I came up with were too long. Suggestions appreciated.

At least it is now commented and typed.

erights replied:
Having filed https://github.com/Agoric/agoric-sdk/issues/9302 , resolving here.

erights commented on packages/async-flow/README.md:
````diff
@@ -0,0 +1,29 @@
+```js
+const outcomeVow = wrapperFunc(...);
+```
+
````
I didn't know about that markdown feature. Cool!

Done, with considerable expansion. PTAL.

mhofman commented:
Note from our conversation earlier:

An invariant violation by the host (e.g. returning a non-passable value, or a promise for now), is not a sign of guest misbehavior. As such the most likely acceptable behavior is to transform the result of the host call into a throw, and record it in the log without failing the flow, so the guest just sees a failure of the host and would handle it "normally".

There is precedent for such behavior in liveslots which transforms promise settlement values into a rejection if the value fails to serialize.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+    },
+    wakeAll() {
+      // [...stuff.keys()] in order to snapshot before iterating
+      const failuresToRestart = [...failures.keys()];
+      const flowsToWake = [...eagerWakers.keys()];
+      for (const flow of failuresToRestart) {
+        flow.restart();
+      }
+      for (const flow of flowsToWake) {
+        flow.wake();
+      }
```
Great question! It turns out that the first answer is yes. And in fact, when I did the https://github.com/Agoric/agoric-sdk/pull/9097#discussion_r1581742043 experiment of deleting the mystery test, I found that when the problem occurred during goodReplay, the flow was in both and the problem happened during the call to `flow.wake()`. Together with this question, it had me looking there for an answer for awhile. But that was not the problem, as explained there.

Turns out you are correct, but not quite as you phrased it. If the flow is in both, `restart()` will clear the failure state, so normally by the time `wake()` happens, it will no longer have failed. But it could have failed again later in that same `restart()` before the `wake()` in which case this loop would call `wake()` on a failed flow.

`wake()` should be idempotent. But in this failure state, `wake()` does
```js
            if (failures.has(flow)) {
              throw failures.get(flow);
            }
```
which isn't technically a non-idempotence of `wake` itself. But it does cause the loop above to exit early for no good reason. Changing to 
```js
            if (failures.has(flow)) {
              return;
            }
```
which worked. But changed again to
```js
            const flowState = flow.getFlowState();
            if (flowState === 'Done' || flowState === 'Failed') {
              return;
            }
```
which also works, with more consistency checking.

erights replied:
So `flow.wake()` should be now be usefully idempotent if `membrane.wake()` is idempotent. I believe it is, so closing. But if I missed something, please reopen.

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,389 @@
+  harden(makeGuestForHostRemotable);
+
+  const makeGuestForHostVow = hVow => {
+    // TODO in order to support E *well*,
+    // use HandledPromise to make `promise` a handled promise for hVow
```
Having captured https://github.com/Agoric/agoric-sdk/issues/9299 for reference in a future PR, I'm resolving here.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.set(flow, fatalProblem);
```
Since exceptions can be thrown internally due to asyncFlow infrastructure failures, I don't want to rule out that we get an infrastructure failure while trying to handle an infrastructure failure. I also don't want to drop an informative diagnostic in favor of a diagnostic about why we couldn't report a diagnostic. However, the code you comment on does not uphold this principle, since it is the first failure likely to be the more important one.

So I changed it to keep the first error as primary but annotate it with the new error. Done.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,417 @@
+   */
+  const asyncFlow = (zone, tag, guestFunc, options = undefined) => {
+    const makeAsyncFlowKit = prepareAsyncFlowKit(zone, tag, guestFunc, options);
+    const hostFuncName = `${guestFunc.name || 'anon'}_hostWrapper`;
```
Good idea! Used (mangled) as only name. The tag is mandatory, so there would be no fallback case. Done

erights commented:
Addressed all review comments except for the one I'm keeping open as I investigate a likely bug. I think this is a narrow enough remaining issue that I can ask y'all to PTAL. Thanks!

erights commented:
Reviewers, do you mind if I do some squashing? (I already got an ok from @mhofman )

erights commented:
All review suggestions so far are now addressed. Whew!

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+              // async IFFE ensures guestResultP is a fresh promise
+              guestAsyncFunc(...guestArgs))();
+
+            if (flow.getFlowState() !== 'Failed') {
+              bijection.init(guestResultP, outcomeKit.vow);
+            }
```
It took me a minute to understand why the state check necessary. Let's add a comment that the bijection would have been reset already in case of a failure.

erights replied:
Done

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+            log.reset();
+            bijection.reset();
+
+            state.isDone = false;
```
Why are we changing `isDone`? I know `restart()` is the only place where this is called from a public path, and it checks that we're not done first, but maybe we should move that check here and never set `isDone` to false (except at init)?

erights replied:
Changed to an assertion. Done.

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+            bijection.reset();
+
+            flow.getFlowState() === 'Failed' ||
+              Fail`Paniced flow must be Failed ${flow}`;
```
```suggestion
              Fail`Panicked flow must be Failed ${flow}`;
```

erights replied:
English is stranger than JS. Done.

mhofman reviewed (changes requested):
The biggest issue I have is with how host vow fulfillment is handled. I think there is a big problem with the current approach, but maybe I missed something. Let's chat.

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+      hVow,
+      async hostFulfillment => {
+        await log.promiseReplayDone();
+        if (guestPromiseMap.get(promise) !== 'settled') {
```
- need to check bijection for this promise
- need to clear guestPromiseMap when we clear ephemeral state
- should avoid redundant watch, but might require more bookkeeping
- analyze / explain fulfilled vow noticed but not recorded because it is during a failed replay
-

erights replied:
Resolved base on verbal discussion w @mhofman

erights replied:
Bug is fixed but still need explanation.

mhofman replied:
The fix I believe hasn't been pushed yet and it requires to check `!stopped` here (and in the rejection handler)

mhofman replied:
Here is the explanation I came up with. Please adapt as you see fit:

> The replay membrane is the only component inserting entries into the log. In particular, the flow's vow durable watcher does not log the settlement outcome, and instead it's the responsibility of the membrane's ephemeral handler. Because of this, the membrane's handler must be careful to:
> - Be added to the vow if the settlement has not yet been recorded in the log
> - Insert a single settlement outcome in the log for the given vow
> 
> In practice the former is accomplished by a handler always being added to the host vow when creating a guest promise, and the handler checking after replay is complete whether the guest promise is already settled (by the log replay) or not. The latter is accomplished by checking that the membrane has not been stopped before updating the log.

erights replied:
Just included it literally. Done.

erights commented on packages/async-flow/README.md:
```diff
@@ -0,0 +1,36 @@
+> [!IMPORTANT]
+> The async function argument should be ***closed***, meaning that it should not use any lexically captured variables other than powerless globals. Any direct access to mutable state or ability to cause effects may introduce bugs, since these effects will happen again under replay outside the control of the asyncFlow isolation and deterministic replay mechanisms.
+>
+> We make an explicit exception for `console`, since log messages sent to `console` are only for diagnostic purposes, and `console` as a whole is write-only. During replay, such out-of-band console log events may appear again. For the same reason, the async function has no obligation to reproduce previous runs of such out-of-band console logging events, since they are outside the replay mechanism ...
```
Expanded to also explain about the loose comparisons of errors. Done.

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+    }
+    const guestReason = hostToGuest(hostReason);
+    status.reject(guestReason);
+    guestPromiseMap.delete(guestPromise);
```
Good catch! But I "fixed" it by setting it to `'settled'`, just like for `doFulfill`. Done

erights commented on packages/async-flow/src/weak-bijection.js:
```diff
@@ -0,0 +1,127 @@
+/**
+ * @param {Zone} zone
+ */
+export const prepareWeakBijection = zone => {
+  /** @type {Ephemera<WeakBijection, VowishStore>} */
+  const g2h = makeEphemera(() => makeVowishStore('guestToHost'));
+  /** @type {Ephemera<WeakBijection, VowishStore>} */
+  const h2g = makeEphemera(() => makeVowishStore('hostToGuest'));
```
> I will file an issue, this can be done as a follow-up

So resolving here.

mhofman replied:
Filed #9365

Wondering if we should not change `WeakMap` to `Map` in the bijection for now. It would have the same memory profile, but would avoid the liveslots overhead of tracking vrefs.

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+    let gRem;
+    /** @param {PropertyKey} [optVerb] */
+    const makeGuestMethod = (optVerb = undefined) => {
+      const guestMethod = (...guestArgs) => {
+        const callIndex = log.getIndex();
+        return guestCallsHost(gRem, optVerb, guestArgs, callIndex);
+      };
```
You're right about it being tricky. I underestimated it. But I found a different simple refactor that seems either elegant or kludgy depending on how you look at it. I'm happy with the technique.

Done.

mhofman replied:
Clever. I wish there was a way to make the temporal aspect more specific, but I don't see anything that's not kludgy.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+          wake() {
+            const { facets } = this;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
```
Ephemera wrong abstraction for this. Switched to ScalarWeakMapStore. Much cleaner. Done.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (isDone) {
```
Ephemera wrong abstraction for this. Switched to ScalarWeakMapStore. Much cleaner. Done.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,453 @@
+            const eph = tmp.for(flow);
+
+            const flowState = flow.getFlowState();
+            if (flowState === 'Done' || flowState === 'Failed') {
```
Done during interactive session w @mhofman

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+  const doFulfill = (hostVow, hostFulfillment) => {
+    const guestPromise = hostToGuest(hostVow);
+    const status = guestPromiseMap.get(guestPromise);
+    if (status === 'settled') {
+      return;
+    }
```
Done interactively w @mhofman

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+      throw panic(fatalError);
+    }
+
+    if (outcome.kind === 'return') {
+      return outcome.result;
+    } else {
+      outcome.kind === 'throw' ||
+        // @ts-expect-error TS correctly knows this case would be outside
+        // the type. But that's what we want to check.
+        Fail`unexpected outcome kind ${q(outcome.kind)}`;
+      throw outcome.problem;
+    }
```
Done

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+      };
+      if (optVerb) {
+        defineProperties(guestMethod, {
+          name: { value: String(optVerb) },
```
Done

erights commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+
+    watchWake(hVow);
+
+    void when(
+      hVow,
```
We believe we (@mhofman and I) fixed this. But needs a test.

erights reviewed (commented):
Github won't let me request changes of myself, but me, please fix this.

mhofman commented on packages/async-flow/src/bijection.js:
```diff
@@ -0,0 +1,132 @@
+      const hostToGuest = h2g.for(self);
+
+      if (guestToHost.has(g)) {
+        guestToHost.get(g) === h ||
```
Oh I totally missed that too!

erights replied:
Done

mhofman commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+  };
+
+  const guestCallsHost = (guestTarget, optVerb, guestArgs, callIndex) => {
+    if (!bijection.hasGuest(guestTarget)) {
```
It might be redundant, but maybe this should check `!stopped` as well ?

erights replied:
Done

mhofman commented on packages/async-flow/src/replay-membrane.js:
```diff
@@ -0,0 +1,393 @@
+    }
+  };
+
+  let stopped = false;
```
We probably should move this earlier in the file

erights replied:
Done

mhofman reviewed (commented):
Audited the membrane for `stopped` state check. I think we could add an extra check in the guest call handler to avoid relying on the bijection being reset.

mhofman commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,491 @@
+  };
+
+  const adminAsyncFlow = outerZone.exo('AdminAsyncFlow', AdminAsyncFlowI, {
+    getFailures() {
+      return failures.snapshot();
+    },
+    wakeAll() {
```
I think these methods should be scoped to a flow definition. In particular it's a mistake to call `wakeAll()` before all `asyncFlow` have been redefined.

IMO we should have the following:
- `prepareAsyncFlowTools` would define an exo class kit for flow definitions. Each instance would hold the `failures` and `eagerWakers` stores in their state, and have the following 2 facets:
  - An "admin facet" would implement the `wakeAll()` and `getFailures()` methods. Maybe it could also implement a `getFlowForOutcomeVow()` method (as a shortcut, we may not need to scope to the definition)
  - An "internal facet" allowing the flow kit behavior methods to interact with the stores of the definition as needed.
- `prepareAsyncFlowTools` would also define a heap `WeakMap` mapping async flow (kit) makers to their respective admin facet of the definition instance.
- `prepareAsyncFlowKit` would use a provide pattern for an instance of the above async flow definition kit. On first definition, an instance would be created and stashed somewhere, either in the provided `zone` (likely using a name derived from the flow's tag), or maybe in a shared store in the `outerZone` (if we're ever interested in easily iterating all flow definitions).
  - Technically the kind of the async flow kit could be held in the state of the definition instance, but I don't believe we can make that happen easily with zones.
  - The behavior methods of the flow kit would close over the "internal facet" of the flow definition to interact with the `failures` and `eagerWakers` stores
- `prepareAsyncFlowKit` would register in the WeakMap the "admin facet" of the flow definition to the maker of the flow kit.
  - it will pin the flow definition kit in memory while the maker is retained, but this whole approach already assumes fairly low cardinality of async flow definitions
- `asyncFlow` would similarly associate the admin facet of the flow definition with the `wrapperFunc`
- Finally, the singleton `adminAsyncFlow` can give access to the flow definition's admin facet through the `WeakMap`

This would also allow to automatically `wakeAll` on the flow's definition (similar to how the flow maker "`restart`s" the flow on instantiation.

erights replied:
With https://github.com/Agoric/agoric-sdk/issues/9375 filed, to be addressed in a later PR, resolving here.

mhofman reviewed (commented):
In the process of reviewing test, I realized the `wakeAll` behavior was not quite as expected it was. Given the change to the data model, I think this needs to be addressed before first release, either now or in a closely following PR.

Furthermore in the upgrading considerations, and maybe somewhere in the documentation, we should mention that because of metering, an activation that executed successfully in a previous incarnation might not replay correctly, even if it doesn't cause any explicit side-effects. That's because metering is a hidden side-effect of any execution.

erights commented on packages/async-flow/src/bijection.js:
```diff
@@ -100,14 +100,14 @@ export const prepareBijection = zone => {
         return true;
       } else {
         !hostToGuest.has(h) ||
-          Fail`internal: unexpected h->g ${h} -> ${guestToHost.get(h)}`;
```
Fixed another bug discovered while testing.

turadg replied:
Are the tests that found these going to be included?

erights replied:
Yes, they are now.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,491 @@
+  });
+
+  // Cannot call this until everything is prepared
+  // adminAsyncFlow.wakeAll();
```
Ok, I just tried uncommenting and got the error suggested by the comment:
```
'no kind info for "17" ("o+d17/1:1"); check deserialize preceding kind definitions',
```
suggestions? Leaving this unfixed and unresolved until we discuss.

mhofman replied:
Yes I think to fix this we'll need to wake up each flow definition separately.

erights replied:
@mhofman verbally suggested doing this in a later turn, so it happens after the first-turn prepares. Done, with note about first turn vs first crank hazard.

erights replied:
Crank hazard now files as https://github.com/Agoric/agoric-sdk/issues/9377

erights replied:
and tested as `async-flow-crank.test.js`

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,491 @@
+  } = outerOptions;
+  const { watch, makeVowKit } = vowTools;
+
+  const failures = outerZone.mapStore('asyncFuncFailures', {
+    keyShape: M.remotable('flow'), // flowState === 'Failed'
+    valueShape: M.error(),
+  });
+
+  const eagerWakers = outerZone.setStore(`asyncFuncEagerWakers`, {
+    keyShape: M.remotable('flow'), // flowState !== 'Done'
+  });
```
Used to be per flow. I changed it to per flow tools on purpose.

erights replied:
With https://github.com/Agoric/agoric-sdk/issues/9375 filed, to be addressed in a later PR, resolving here.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,491 @@
+            admin.reset();
+            if (eager) {
+              eagerWakers.add(flow);
+            }
```
The eagerSet is a SetStore which already consolidates redundant `add`s. I did add the case of removing it if !eager and already there. Done.

erights commented on packages/async-flow/src/async-flow.js:
```diff
@@ -0,0 +1,491 @@
+            // console logging and
+            // resource exhaustion, including infinite loops
+            const err = makeError(
+              X`In a Failed state: see getFailures() for more information`,
```
Done

erights commented on packages/async-flow/src/convert.js:
```diff
@@ -0,0 +1,131 @@
+        }
+        const tag = getTag(specimen);
+        const { payload } = specimen;
+        return makeTagged(tag, convertRecur(payload, `${tag} payload`));
```
Genuine bug was that I forgot to convert the payload, which was a cross membrane capability leak!

Found while looking at the code for other reasons. Fixed.

mhofman replied:
And I missed it in review...

erights commented on packages/async-flow/src/convert.js:
```diff
@@ -0,0 +1,131 @@
+        if (isVow(specimen)) {
+          return convertPromiseOrVow(specimen);
+        }
+        const tag = getTag(specimen);
```
Genuine bug I was trying to get the tag with `specimen.tag`, rather than the `getTag` function. Found by testing. Fixed.

erights commented on packages/async-flow/src/equate.js:
```diff
@@ -69,7 +69,7 @@ export const makeEquate = bijection => {
       }
       case 'tagged': {
-        equate(g.tag, h.tag, 'tag');
+        equate(getTag(g), getTag(h), 'tag');
```
Genuine bug I was trying to get the tag with specimen.tag, rather than the getTag function. Found similar bug in convert by testing, and then this by looking for other occurrences. Not caught by testing because `undefined === undefined`. Fixed.

mhofman commented on packages/async-flow/test/replay-membrane-settlement.test.js:
```diff
@@ -0,0 +1,149 @@
+  await eventLoopIteration();
+
+  const p1 = mem.hostToGuest(v1);
+  mem.wake();
```
Maybe check that the log hasn't changed (no entry of p2 resolution) until the membrane is awoken? Could be done by moving the `log.dump()` check here?

mhofman reviewed (approved):
The tests look mostly good. I'm a little concerned that the waking behavior on new incarnation is not quite tested right, in particular if a flow is not failed and not eagerly waking, whether the durable vow watcher kicks in.

Conditional approval on tweaking this if easily doable (I think it should be now that we fixed auto waking behavior)

erights commented on packages/async-flow/test/bad-host.test.js:
```diff
@@ -0,0 +1,203 @@
+  const outcomeV = zone.makeOnce('outcomeV', () => Fail`need outcomeV`);
+
+  // TODO I shouldn't need to do this.
+  await adminAsyncFlow.wakeAll();
```
Gone

erights commented on packages/async-flow/test/convert.test.js:
```diff
@@ -0,0 +1,127 @@
+  t.deepEqual(hostToGuest(URIError('msg')), URIError('msg'));
+
+  t.deepEqual(guestToHost({ g1, g2 }), { g1: h1, g2: h2 });
+  t.deepEqual(hostToGuest({ h1, h2 }), { h1: g1, h2: g2 });
```
Done

erights commented on packages/async-flow/test/convert.test.js:
```diff
@@ -0,0 +1,127 @@
+  t.not(g1, g2);
+  t.is(hostToGuest(h1), g1);
+  t.is(hostToGuest(h2), g2);
+  t.is(hostToGuest(h3), g3);
```
I'm curious to see how this plays out. In any case, sounds like I can resolve here.

erights commented on packages/async-flow/test/equate.test.js:
```diff
@@ -0,0 +1,116 @@
+      'cannot yet send guest remotables to host "[Alleged: g1]" vs "[Alleged: h1]"',
+  });
+  bij.init(g1, h1);
+  equate(g1, h1);
```
Done

erights commented on packages/async-flow/test/bijection.test.js:
```diff
@@ -0,0 +1,84 @@
+ * @param {any} t
+ * @param {Zone} zone
+ */
+const testBijection = (t, zone) => {
```
Done

erights commented on packages/async-flow/test/bad-host.test.js:
```diff
@@ -0,0 +1,203 @@
+      t.log('  badHost replay1 to hang awaiting p2');
+      // awaiting a promise that won't be resolved until next incarnation
+      await p1;
+      t.fail('must not reach here in replay 1');
```
> re the log message, looks like we're awaiting p1, not p2

This part done so far.

erights commented on packages/async-flow/test/async-flow.test.js:
```diff
@@ -0,0 +1,365 @@
+  // the next incarnation.
+  hOrch7.resolve('y');
+  // TODO I shouldn't need to do this.
+  await adminAsyncFlow.wakeAll();
```
Done

erights commented on packages/async-flow/test/async-flow.test.js:
```diff
@@ -0,0 +1,365 @@
+
+  const { guestMethod } = {
+    async guestMethod(_gOrch7, _gP, _p3) {
+      t.fail('Must not replay this');
```
Done

erights commented on packages/async-flow/test/async-flow.test.js:
```diff
@@ -0,0 +1,365 @@
+      t.log('  goodReplay about to await p2');
+      // awaiting a promise that won't be resolved until this incarnation
+      await p2;
+      t.log('  goodReplay woke up!');
```
Filing https://github.com/Agoric/agoric-sdk/issues/9383 to remember that this still needs to be addressed after this PR.

Resolving.
